### ISSUE TITLE
reasonable frequency to sample events

### ISSUE THREAD

**Issue by adelavega** [neuroscout/neuroscout#834]
Currently, we've sampled events at somewhat inconsistent frequencies. 

For example, visual events (scene labels, face detection) are mostly sampled at 1hz, because we made the arbitrary choice to sample frames at 1hz from the video (i.e. one single frame was chosen from each 1 second window, and saved).

For auditory events, the sampling frequency varies depending on the extractor, and can be somewhat inconsistent, but is usually higher than 1hz. 

@jsmentch recently has been computing regressors that are dependent on visual and auditory events co-occurring, and therefore this mismatch made things more difficult. In addition, he thought that 1hz was too low, as he might miss a lot of events for this type of measure. 

I want to think again about what a "good enough" frequency for events in Neuroscout is.

@satra previously suggested 10-20hz.

however, there are two major downsides: 1) it's expensive for paid extractors (which to be fair we are going away from) 2) it increases how long db queries take, in general just makes things more slow and unwieldy 3) takes up a lot of disk space (probably solvable but the database has grown substantially (were at 20GB currently). 

one potential solution is to sample at 10hz and then downsample to 1-2hz, when storing in the database (using a Pliers downsample filter). this should at least represent the average of the sampled frame more accurately (versus one random frame). However, this "averaged" sample would not help with @jsmentch's logical operations (e.g. "face_1" and "speaker_1"). 

another alternative is to upsample slightly to 2-3hz or so, and say that's good enough. 

it's not clear how to validate aside from subjective judgments when looking at timecourses of the features. 
some would argue we're splitting hairs and with the low temporal resolution of fMRI it won't matter much anyways.

one additional question we had was that if we were to downsample from 20hz to 1hz (for example), should the onset of this event be at 0s, or 0.5s (the middle of the frame)? and would this decision be affected the slice timing correction?

any additional thoughts, opinions, requests? @satra @tyarkoni @effigies

**Comment by satra** [comment]
> 3) takes up a lot of disk space

@adelavega - is storage an issue because a TSV file is being used? at the very least TSV.gz could be used. but perhaps something that allows for sparse + dense timeseries storage would be most appropriate for extracted features. (see https://github.com/TileDB-Inc/TileDB as an example).

ideal scenario: for me any dense extracted features (e.g., intensity/loudness, prob of being inside vs outside) should be stored at the audio or video resolution of the original movie data. but i suspect most extracted features can be represented as or transformed into sparse event+duration timeseries. 

so i think not all things should be dense, and perhaps that's sufficient to address the sampling question for 90% of features. and any dense feature can be converted to non-dense by a threshold operation. so a practical consideration is to suggest that what can be thresholded easily should be thresholded. 

this is for future consideration: there is another potential type of dataformat which mixes sparse + dense. for example: extract timeseries motifs whether from video or audio at certain event triggers. (the closest neuroimaging equivalent would be ERP style designs). the triggers are sparse, but the data around the triggers are stored as dense/sparse. this is more about a hierarchical descriptor for a time series.

**Comment by tyarkoni** [comment]
I think we should start by taking one or two 5-minute clips, and for a few different features, plot predictor timecourses at:
1. 1 Hz discrete sampling rate (i.e., current approach)
2. 10 Hz sampling rate
3. 10 Hz sampling rate downsampled to 1 Hz

(3) shouldn't be much more cumbersome to extract than (2) (though it will likely require a shift to pretrained models everywhere, which as you say is already the plan anyway), and should definitely be preferable to (1). I think the bar for going with (2) over (3) would be pretty high given the space/time constraints you mention. But conceivably we could go to 2 or 3 Hz, which would also be good to plot.

This would still preclude the kinds of fine-grained operations suggested by @jsmentch, which would be unfortunate, but I don't see that as being a major use case for the NeuroScout platform. Probably the way to handle that would be to develop a separate series of workflows that wrap pliers/pybids/fitlins and extract features directly into the design matrix (i.e., a greatly simplified version of what you have right now, @adelavega, that completely avoids all the databasing steps). Then NeuroScout becomes the entrance-level tool for people who need a GUI, and more powerful operations can be done with the programmatic tools.

**Comment by tyarkoni** [comment]
@satra the issue is that all the events are represented in a relational DB, and aside from the storage issues, building design matrices on-the-fly via the web for models that could involve dozens of predictors, each sampled at high frequency, isn't really our idea of a good time... we want to keep the backend streamlined and maintainable, and to keep operating costs low down the line. In practice, NeuroScout isn't really aimed at people who want to do analyses at 30 Hz resolution. But, as noted above, I'd be happy to work on a separate set of tools that more efficiently construct complex design matrices without all the NeuroScout overhead.

**Comment by satra** [comment]
@tyarkoni - the separation sounds good to me with regard to the relation DB. 

however, from a reusability perspective, does it make sense to rerun the extractors across everyone. could we store the outputs of the pliers extractors for reuse outside the db, perhaps in a separate object that can handle differential temporal structures? we could maintain this outside of neuroscout and neuroscout could import or transform this for GUI based use cases.

**Comment by satra** [comment]
oh, it may be good to create a couple of diagrams the illustrate the web vs programmatic workflows and use cases.

**Comment by tyarkoni** [comment]
@satra in principle, storage of extracted features could be centralized, sure. In practice, I don't think we have the bandwidth to build such a thing right now (and likely won't before the grant ends). I also suspect that the actual rate of re-use, as opposed to novel extraction, would be quite low. Presumably stored data would have to be tied to specific versions of the stimuli, pre-trained models, extractors, and their parameters, and every small change would necessitate re-extracting everything (unless we built a lot of logic into the back end, at which point it becomes a big project in its own right). I imagine we would just end up with a rapidly ballooning DB that's almost entirely writes, with very few reads. So I think bang for the buck is higher if we were to focus on the workflow part—at least for the time being.

**Comment by adelavega** [comment]
@tyarkoni to add to your separate workflow, we could also make it easy for those predictors to be contributed back to neuroscout via the API (with force downsampling upon ingestion if they're too fine).

I personally think 10hz downsampled to 2-3hz might be a sweet spot, as we already have some features at higher sampling rates without performance issues (although it's a linear relationship with query time, so it will be slower than 1hz).

@satra in terms of sparse/dense, all events are "sparse". zero values are not explicitly stored. this works well except for inherently dense predictors. I think a light thresholding could help in a lot of cases (i.e. no point in storing a value that's almost zero but not quite), so thanks for the suggestion.

i'll also throw in that if specific features really needed higher sampling rates, we could allow it on a case by case basis, just not as a default across the board.

**Comment by adelavega** [comment]
Closed with #702 , downsampling from original sampling rate to 3hz where necessary. for visual features, there is still a minor implementational hurdle to do this, and we may end up simply extracting at 3hz, rather than 10hz-3hz (in practice, results were identical with auditory test features).
